File: module_wan/src/main/java/com/huyingbao/module/wan/ui/view/ArticleListFragment.java
package com.huyingbao.module.wan.ui.view;

import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.listener.OnItemClickListener;
import com.huyingbao.core.arch.model.RxChange;
import com.huyingbao.core.arch.scope.ActivityScope;
import com.huyingbao.core.common.R2;
import com.huyingbao.core.common.view.CommonRxFragment;
import com.huyingbao.core.common.widget.CommonLoadMoreView;
import com.huyingbao.module.wan.R;
import com.huyingbao.module.wan.ui.action.WanAction;
import com.huyingbao.module.wan.ui.action.WanActionCreator;
import com.huyingbao.module.wan.ui.adapter.WanAdapter;
import com.huyingbao.module.wan.ui.model.Article;
import com.huyingbao.module.wan.ui.store.WanStore;

import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;

/**
 * Created by liujunfeng on 2017/12/7.
 */
@ActivityScope
public class ArticleListFragment extends CommonRxFragment<WanStore> {
    @Inject
    WanActionCreator mActionCreator;
    @BindView(R2.id.rv_content)
    RecyclerView mRvContent;
    @BindView(R2.id.srl_content)
    SwipeRefreshLayout mSrlContent;

    private List<Article> mDataList;
    private BaseQuickAdapter mAdapter;

    private int mNextRequestPage = 1;
    private static final int PAGE_SIZE = 20;

    @Inject
    public ArticleListFragment() {
    }

    @Nullable
    @Override
    public WanStore getRxStore() {
        return ViewModelProviders.of(getActivity(), mViewModelFactory).get(WanStore.class);
    }

    @Override
    public int getLayoutId() {
        return R.layout.common_fragment_base_list;
    }

    @Override
    public void afterCreate(Bundle savedInstanceState) {
        initActionBar("仓库列表");
        initRecyclerView();
        initAdapter();
        //addHeadView();
        initRefreshLayout();
        showData();
        //如果store已经创建并获取到数据，说明是横屏等操作导致的Fragment重建，不需要重新获取数据
        if (getRxStore().isCreated()) return;
        mSrlContent.setRefreshing(true);
        refresh();
    }

    /**
     * 接收RxChange，粘性
     */
    @Override
    @Subscribe(sticky = true)
    public void onRxChanged(@NonNull RxChange rxChange) {
        super.onRxChanged(rxChange);
        switch (rxChange.getTag()) {
            case WanAction.GET_ARTICLE_LIST:
//                startActivity(RandomActivity.newIntent(this, getRxStore().getCategory()));
                break;
        }
    }

    /**
     * 实例化RecyclerView
     */
    private void initRecyclerView() {
        mRvContent.setLayoutManager(new LinearLayoutManager(getActivity()));
        mRvContent.setHasFixedSize(true);
        mRvContent.setLayerType(View.LAYER_TYPE_SOFTWARE, null);//硬件加速
        mRvContent.addOnItemTouchListener(new OnItemClickListener() {
            @Override
            public void onSimpleItemClick(BaseQuickAdapter adapter, View view, int position) {
            }
        });
    }

    /**
     * 实例化adapter
     */
    private void initAdapter() {
        mDataList = new ArrayList();
        mAdapter = new WanAdapter(mDataList);
        //设置加载更多监听器
        mAdapter.setOnLoadMoreListener(() -> loadMore(), mRvContent);
        //设置加载动画
        mAdapter.openLoadAnimation(BaseQuickAdapter.SLIDEIN_LEFT);
        //view设置适配器
        mRvContent.setAdapter(mAdapter);
    }

    /**
     * 添加头部view
     */
    private void addHeadView() {
        View headView = getLayoutInflater().inflate(R.layout.common_view_head, (ViewGroup) mRvContent.getParent(), false);
        headView.findViewById(R.id.iv_head).setVisibility(View.GONE);
        ((TextView) headView.findViewById(R.id.tv_head)).setText("change load view");
        headView.setOnClickListener(v -> {
            mAdapter.setNewData(null);
            mAdapter.setLoadMoreView(new CommonLoadMoreView());
            mRvContent.setAdapter(mAdapter);
            Toast.makeText(getActivity(), "change complete", Toast.LENGTH_LONG).show();
            mSrlContent.setRefreshing(true);
            refresh();
        });
        mAdapter.addHeaderView(headView);
    }

    /**
     * 实例化view
     */
    private void initRefreshLayout() {
        mSrlContent.setColorSchemeColors(Color.rgb(47, 223, 189));
        mSrlContent.setOnRefreshListener(() -> refresh());
    }

    /**
     * 显示数据
     */
    private void showData() {
        getRxStore().getGitRepoList().observe(this, products -> {
            if (products == null) return;
            //判断获取回来的数据是否是刷新的数据
            boolean isRefresh = mNextRequestPage == 1;
            setData(isRefresh, products.getData().getDatas());
            mAdapter.setEnableLoadMore(true);
            mSrlContent.setRefreshing(false);
        });
    }

    /**
     * 刷新
     */
    private void refresh() {
        mNextRequestPage = 1;
        mAdapter.setEnableLoadMore(false);//这里的作用是防止下拉刷新的时候还可以上拉加载
        mActionCreator.getArticleList(mNextRequestPage);
    }

    /**
     * 加载更多
     */
    private void loadMore() {
        mActionCreator.getArticleList(mNextRequestPage);
    }

    /**
     * 设置数据
     *
     * @param isRefresh
     * @param data
     */
    private void setData(boolean isRefresh, List<Article> data) {
        mNextRequestPage++;
        final int size = data == null
                ? 0
                : data.size();
        if (isRefresh) {
            mAdapter.setNewData(data);
        } else {
            if (size > 0) {
                mAdapter.addData(data);
            }
        }
        if (size < PAGE_SIZE) {
            //第一页如果不够一页就不显示没有更多数据布局
            mAdapter.loadMoreEnd(isRefresh);
            Toast.makeText(getContext(), "no more data", Toast.LENGTH_SHORT).show();
        } else {
            mAdapter.loadMoreComplete();
        }
    }
}

File: module-wan/src/main/java/com/huyingbao/module/wan/ui/article/view/ArticleListFragment.kt
package com.huyingbao.module.wan.ui.article.view

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.android.arouter.launcher.ARouter
import com.huyingbao.core.arch.model.RxLoading
import com.huyingbao.core.base.flux.fragment.BaseFluxFragment
import com.huyingbao.module.common.app.CommonAppStore
import com.huyingbao.module.wan.R
import com.huyingbao.module.wan.ui.article.action.ArticleAction
import com.huyingbao.module.wan.ui.article.action.ArticleActionCreator
import com.huyingbao.module.wan.ui.article.adapter.ArticleAdapter
import com.huyingbao.module.wan.ui.article.store.ArticleStore
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import org.greenrobot.eventbus.Subscribe
import org.jetbrains.anko.find
import javax.inject.Inject

/**
 * Created by liujunfeng on 2019/1/1.
 */
class ArticleListFragment : BaseFluxFragment<ArticleStore>() {
    @Inject
    lateinit var articleActionCreator: ArticleActionCreator

    private var articleAdapter: ArticleAdapter? = null
    private var refreshLayout: SmartRefreshLayout? = null

    companion object {
        fun newInstance(): ArticleListFragment {
            return ArticleListFragment()
        }
    }

    override fun getLayoutId(): Int {
        return R.layout.common_fragment_list
    }

    override fun afterCreate(savedInstanceState: Bundle?) {
        setTitle(R.string.wan_label_article, true)
        initAdapter()
        initRefreshView()
    }

    /**
     * 设置Adapter
     */
    private fun initAdapter() {
        //实例化adapter
        articleAdapter = ArticleAdapter()
        //RecyclerView设置适配器
        view?.find<RecyclerView>(R.id.rv_content)?.adapter = articleAdapter
        //显示数据
        rxStore?.articleLiveData?.observe(this, Observer {
            articleAdapter?.submitList(it)
        })
    }

    /**
     * 初始化上下拉刷新View
     */
    private fun initRefreshView() {
        refreshLayout = view?.find(R.id.rfl_content)
        refreshLayout?.setOnRefreshListener {
            articleActionCreator.getArticleList(0)
        }
        //如果是新创建，调用刷新方法，排除屏幕旋转
        if (rxStore?.nextRequestPage == 0) {
            refreshLayout?.autoRefresh()
        }
    }

    /**
     * 显示进度对话框
     * 接收[RxLoading]，粘性
     * 该方法不经过RxStore,
     * 由RxFluxView直接处理
     */
    @Subscribe(sticky = true)
    fun onRxLoading(rxLoading: RxLoading) {
        if (!rxLoading.isLoading) {
            refreshLayout?.finishRefresh()
        }
    }

    /**
     * fragment中创建menu
     */
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        //Menu布局文件名同界面布局文件名
        inflater.inflate(R.menu.wan_fragment_article_list, menu)
    }

    /**
     * menu点击事件
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when {
            item.itemId == R.id.menu_to_banner -> {
                //跳转BannerFragment
                articleActionCreator.postLocalChange(ArticleAction.TO_BANNER)
                true
            }
            item.itemId == R.id.menu_to_friend -> {
                //跳转FriendFragment
                articleActionCreator.postLocalChange(ArticleAction.TO_FRIEND)
                true
            }
            item.itemId == R.id.menu_to_gan -> {
                //跳转module-gan
                ARouter.getInstance().build(CommonAppStore.RandomActivity).navigation()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
